src/ocrd/processor/base.py

```diff
@@ -206,22 +206,23 @@ def ocrd_tool(self) -> dict:
         return self.metadata['tools'][self.executable]
 
     @property
-    def parameter(self) -> Optional[dict]:
+    def parameter(self) -> frozendict[str, Any]:
         """the runtime parameter dict to be used by this processor"""
         if hasattr(self, '_parameter'):
             return self._parameter
-        return None
+        return frozendict({})
 
     @parameter.setter
     def parameter(self, parameter : dict) -> None:
-        if self.parameter is not None:
+        if hasattr(self, '_parameter'):
             self.shutdown()
+            delattr(self, '_parameter')
```

Comment:
I don't see any benefits of this.

We cannot use `cached_property`, because our setter is more complex. But the hidden attribute serves us the `cached` behaviour just fine. The `None` state I find more clear than the `not hasattr(self, '_parameter')`.

The other changes are fine, though. In order not to put another break on `b1` (after merging #14) – could you please cherry-pick / rebase this and post a new PR?

Reply:
> I don't see any benefits of this.

The benefit is that `@property parameter` will always return a `(frozen)dict`, so we don't need to have a `-> Optional[dict]` with `assert self.parameter` later on to convince type checking that `self.parameter` is, at this point definitely not None.

> The `None` state I find more clear than the `not hasattr(self, '_parameter')`.

I would have agreed before having pyright in my editor showed me how many `Optional[type]` there are in the codebase ;)

> could you please cherry-pick / rebase this and post a new PR?

You mean with just the comments and typing added but without the change in `self.parameter`?

Reply:
Ok, so the reasoning is that `Optional` is ugly (at least for some checkers/editors).

We currently use `None` for the uninitialized case. Returning an empty dict for this is counterintuitive: There may be required parameters, and there may be defaults to be expanded by the validator. In contrast, where we actually do need/expect parameters, we can assert/annotate `dict`.

Exchanging `None` with `hasattr(self, '_parameter')` is even more ugly IMHO. We only introduced `_parameter` for the caching of the validator result.

But if you insist... I'll go through the existing tests and v3 ports to see what the ramifications of that might be, and if everything is fine, we can move v3 to that.

> > could you please cherry-pick / rebase this and post a new PR?
> 
> You mean with just the comments and typing added but without the change in `self.parameter`?

Yes, for the case where we agree to keep `None` – the other changes are good, but come in the same single commit.